src/iologindata.cpp

```diff
@@ -52,12 +52,40 @@ bool IOLoginData::saveAccount(const Account& acc)
 	return Database::getInstance()->executeQuery(query.str());
 }
 
+std::string decodeSecret(const std::string& secret) {
+	// simple base32 decoding
+	std::string key;
+	key.reserve(10);
+
+	int buffer = 0, left = 0;
+	for (const auto& ch : secret) {
+		buffer <<= 5;
+		if (ch >= 'A' && ch <= 'Z') {
+			buffer |= (ch & 0x1F) - 1;
+		} else if (ch >= '2' && ch <= '7') {
+			buffer |= ch - 24;
+		} else {
+			// if a key is broken, return empty and the comparison
+			// will always be false since the token must not be empty
+			return {};
+		}
+
+		left += 5;
+		if (left >= 8) {
+			left -= 8;
+			key.push_back(static_cast<char>(buffer >> left));
```

Comment:
I know I'm being annoyingly pedantic, but right shift on a signed integer has implementation-defined behavior. You're also making assumptions about bit-length of int which is implementation defined as well.